### PR TITLE
175 filtering using auxscore

### LINKED ISSUE #175: how to apply filters
In UI user can define filters. We could add them in Fluka at least for Fluence scoring.

Fluence -> thats easy, you specify generalized particle.
For dose -> use AUXSCORE ?

When using AUXSCORE you can check if in the JSON coming from UI you have Z and A defined.

image
image

### PR DESCRIPTION
**Pull request by hendzeld**
Resolves #175

### CODE REVIEW

File: converter/fluka/cards/scoring_card.py [hunk 1]
```diff
@@ -1,42 +1,112 @@
...
-from converter.fluka.helper_parsers.scoring_parser import Scoring
+from converter.fluka.helper_parsers.detector_parser import Detector
+from converter.fluka.helper_parsers.scoring_parser import CustomFilter, ParticleFilter, Quantity, Scoring
 
 
-def handle_first_card(scoring: Scoring, output_unit: int = 21) -> str:
-    """Creates first Scoring card"""
+@dataclass
+class ScoringCardIndexCounter:
+    """Class representing counter for Fluka scoring cards"""
```
Comment (grzanka) [inline converter/fluka/cards/scoring_card.py:11]: Why do you need a dedicated class here ? simple int would be enough

File: converter/fluka/cards/scoring_card.py [hunk 2]
```diff
@@ -1,42 +1,112 @@
...
+    output: list[str] = []
+    for quantity in scoring.quantities:
+        usrbin_card = handle_usrbin_scoring(scoring.detector, quantity, output_unit, counter)
+        output.append(usrbin_card)
+    return '\n'.join(output).strip()
+
+
+def short_name(name: str) -> str:
+    """Creates short name for fluka card"""
+    return f'{name[:10]}'
```
Comment (grzanka) [inline converter/fluka/cards/scoring_card.py:27]:
```suggestion
    return name[:10]
```

Add a comment explaining why only 10 characters are taken.

File: converter/fluka/cards/scoring_card.py [hunk 3]
```diff
@@ -1,42 +1,112 @@
...
-    first_card.what = [
-        binning_what, particle_of_scoring, output_unit, scoring.detector.x_max, scoring.detector.y_max,
-        scoring.detector.z_max
-    ]
-    first_card.sdum = 'changeme'
+    if not quantity_to_score:
+        return ''
 
-    return first_card.__str__()
+    output_unit = str(output_unit * -1)
```
Comment (grzanka) [inline converter/fluka/cards/scoring_card.py:53]: Add a comment explaining what is going on here.
Its bad practice to modify parameter passed to the function, imagine what would happen if the function is executed twice.

Best would be to create new variables, say `output_unit_in_fluka_convention` or something meaningfluf

File: converter/fluka/cards/scoring_card.py [hunk 4]
```diff
@@ -1,42 +1,112 @@
...
+    return f'{auxscore!s}'
+
+
+def parse_filter_value(scoring_filter: Union[CustomFilter, ParticleFilter]) -> Optional[Union[int, str]]:
+    """Parses filter value from filter"""
+    if isinstance(scoring_filter, ParticleFilter):
+        return scoring_filter.particle
+    if isinstance(scoring_filter, CustomFilter):
+        scoring_filter: CustomFilter
+        return -(scoring_filter.z * 100 + scoring_filter.a * 100000)
```
Comment (grzanka) [inline converter/fluka/cards/scoring_card.py:107]: Add a link to Fluka documentation which explains the rule

File: converter/fluka/helper_parsers/scoring_parser.py [hunk 5]
```diff
@@ -1,36 +1,148 @@
...
+    return ParticleFilter(name=particle['name'], particle=_particle_mappings[particle['name']])
+
+
+def get_custom_filter(filter_dict: dict) -> list[CustomFilter]:
+    """Creates CustomFilter from dictionary.
+
+    Returns None if filter cannot be created for Fluka.
+    """
+    if not filter_dict.get('rules'):
+        return None
```
Comment (grzanka) [inline converter/fluka/helper_parsers/scoring_parser.py:83]:
```suggestion
        return []
```

return empty list or adjust type hints for the output

File: workspace/info.json [hunk 6]
```diff
@@ -0,0 +1 @@
+{'version': '', 'label': '', 'simulator': 'shieldhit'}
```
Comment (grzanka) [inline workspace/info.json:1]: do we need that file ?

File: converter/fluka/cards/scoring_card.py (hunk 1 above)
Comment (hendzeld) [reply converter/fluka/cards/scoring_card.py:11]: We can have multiple scorings with multiple quantities, therefore we have to use wrapper (class).
Which will be shared by all scorings.
We are alos incrementing this counter inside of `handle_scoring_cards` function.
And sometimes we do not increment its value.

In future, it should be moved to class responsible for parsing all types of scorings etc.

File: converter/fluka/helper_parsers/scoring_parser.py (hunk 5 above)
Comment (hendzeld) [reply converter/fluka/helper_parsers/scoring_parser.py:83]: Type hint was misleading, function returns None or CustomFilter.
I forgot to update it after the refactoring.

**Comment by grzanka** [comment]
Integration tests on 270 MeV/u carbon ion beam

Expected results on fluence and dose:
image
image

**Comment by grzanka** [comment]
Plots from https://arxiv.org/pdf/1410.1378.pdf

**Comment by hendzeld** [comment]
[24_01_05_run7.zip](https://github.com/yaptide/converter/files/13845549/24_01_05_run7.zip)

Project + Generated Fluka input and results from pymchelper

**Review by grzanka — APPROVED** [review]

**Pull request merged**